agg-ast/schema_derivation/src/schema_derivation.rs

```diff
@@ -149,6 +169,184 @@ impl DeriveSchema for Stage {
+                LookupFrom::Collection(ref c) => Namespace(state.current_db.clone(), c.clone()),
+                LookupFrom::Namespace(ref n) => Namespace(n.db.clone(), n.coll.clone()),
+            }
+        }
+
+        fn derive_equality_lookup_schema(
+            lookup: &EqualityLookup,
+            state: &mut ResultSetState,
+        ) -> Result<Schema> {
+            let from_ns = from_to_ns(&lookup.from, state);
+            let from_schema = state
+                .catalog
+                .get(&from_ns)
+                .ok_or_else(|| Error::UnknownReference(from_ns.into()))?;
+            insert_required_key_into_document(
+                &mut state.result_set_schema,
+                Schema::Array(Box::new(from_schema.clone())),
+                vec![lookup.as_var.to_string()],
+            );
+            Ok(state.result_set_schema.to_owned())
+        }
+
+        fn derive_concise_lookup_schema(
+            lookup: &ConciseSubqueryLookup,
+            state: &mut ResultSetState,
+        ) -> Result<Schema> {
+            derive_subquery_lookup_schema_helper(
+                &lookup.let_body,
+                &lookup.from,
+                &lookup.pipeline,
+                &lookup.as_var,
+                state,
+            )
+        }
+
+        fn derive_subquery_lookup_schema(
+            lookup: &SubqueryLookup,
+            state: &mut ResultSetState,
+        ) -> Result<Schema> {
+            derive_subquery_lookup_schema_helper(
+                &lookup.let_body,
+                &lookup.from,
+                &lookup.pipeline,
+                &lookup.as_var,
+                state,
+            )
+        }
+
+        fn derive_subquery_lookup_schema_helper(
+            // generally, we do not pass &Option<Type> but usually Option<&Type>, but this is an
+            // internal helper function.
+            let_body: &Option<LinkedHashMap<String, Expression>>,
+            from: &Option<LookupFrom>,
+            pipeline: &[Stage],
+            as_var: &str,
+            state: &mut ResultSetState,
+        ) -> Result<Schema> {
+            // the output of the lookup stage is the current result with the result of the pipeline
+            // nested under the as_var. We can derive the schema for the from collection, and then
+            // apply the pipeline to that schema to get the result.
+            let mut variables = let_body
+                .iter()
+                .flatten()
+                .map(|(k, v)| Ok((k.clone(), v.derive_schema(state)?)))
+                .collect::<Result<BTreeMap<String, Schema>>>()?;
+            let from_ns = from_to_ns(from.as_ref().ok_or(Error::MissingFromField)?, state);
+            let from_schema = state
+                .catalog
+                .get(&from_ns)
+                .ok_or_else(|| Error::UnknownReference(from_ns.into()))?;
+            variables.extend(state.variables.clone());
+            let mut lookup_state = ResultSetState {
```

Comment:
really cool stuff here, creating a new result set state. I imagine this is one of the only times well do this as opposed to cloning the original